src/FastSerialization/SegmentedList.cs

```diff
@@ -403,10 +421,28 @@ public T[] ToArray()
         /// <param name="arrayIndex">Destination array starting index.</param>
         public void CopyTo(T[] array, int arrayIndex)
         {
-            int remain = this.count;
+            if (array == null)
+            {
+                throw new ArgumentNullException(nameof(array));
+            }
+
+            if (arrayIndex < 0 || arrayIndex > array.Length)
```

Comment:
I believe this should be `arrayIndex >= array.Length`, and the argument text updated.